<div align=center><img src="https://logosdownload.com/logo/national-university-of-singapore--logo-big.png" width="30%" height="50%"></div>



<center><font face="Time new roman" size = 10 align=center>EE5103 CA 2</font> </center>
<center><font face="Time new roman" size = 4 align=center>Guo Shiping</font> </center>
<center><font face="Time new roman" size = 4 align=center> A0260014Y</font> </center>

<font face="Time new roman" size = 7 align=center bold>Q1</font> 

<font face="Time new roman" size = 6 align=center bold>a)</font> 


$$ Y(s) =  
\frac{1}{s(Ts+1)} U(s) = \frac{10}{s^2+10s}U(s) $$ 
$$ \ddot{y} = -10\dot{y} + 10u$$
$$ \because x(t) = (\dot{y(t)}, y(t))^T$$
$$ \therefore The\ continuous-time\ state-space\ model\ is:$$
$$
\dot{\begin{bmatrix}
x_1 \\
x_2 
\end{bmatrix}} = {\begin{bmatrix}
-10 & 0\\
1  & 0
\end{bmatrix}} {\begin{bmatrix}
x_1 \\
x_2 
\end{bmatrix}}+ {\begin{bmatrix}
10 \\
0 
\end{bmatrix}}u$$
$$ y = [0, 1] 
{\begin{bmatrix}
x_1 \\
x_2 
\end{bmatrix}}
$$
$$ A = {\begin{bmatrix}
-10 & 0\\
1  & 0
\end{bmatrix}},  B = {\begin{bmatrix}
10 \\
0
\end{bmatrix}}, C = [0, 1]$$
$After\ sampling,\ the\ state\ space\ changes\ into\ discrete-time\ form:$
$$x(k+1) = \Phi x(k) + \Gamma u(k) $$
$$\Phi = e^{Ah},  \Gamma =\int_{0}^{h}e^{Av}dv  $$

<font face="Time new roman" size = 6 align=center bold>b)</font> 

$$\lambda_0 = -10, \lambda_1 = 0 $$
$$h(\lambda) = \beta_0 + \beta_1 \lambda $$
$$ \left\{
\begin{aligned}
e^{\lambda_0h} & = & \beta_0 + \beta_1 \lambda_0 \\
e^{\lambda_1h} & = & \beta_0 + \beta_1 \lambda_1 
\end{aligned} \right.
=\left\{
\begin{aligned}
e^{-10h} & = & \beta_0 - 10\beta_1  \\
e^{0h} & = & \beta_0 
\end{aligned}
\right. $$
$$\left\{
    \begin{matrix} 
    \beta_0 & = & 1  \\ 
    \beta_1 & = & \frac {1-e^{-10h}}{10}
    \end{matrix}
    \right.
$$
$$\Phi = e^{Ah} = \beta_0 I+ \beta_1 A = {\begin{bmatrix}
e^{-10h} & 0\\
\frac{1-e^{-10h}}{10}  & 1
\end{bmatrix}} $$
$$ \Gamma =\int_{0}^{h}e^{Av}dvB = \int_0^h  {\begin{bmatrix}
e^{-10v} & 0\\
\frac{1-e^{-10v}}{10}  & 1
\end{bmatrix}} dv 
{\begin{bmatrix}
10\\
0
\end{bmatrix}} = {\begin{bmatrix}
1-\frac{1}{10}e^{-10h}\\
h+\frac{1}{10}(-1+e^{-10h}) 
\end{bmatrix}} $$


$Assuming\ that\  u(k) = -Lx(k), using\ the\ deadbeat\ controller:$
$Use\ Ackermann's\ formula$
$$A_m(\phi) = \phi^2 + p_1\phi + p_2I \ \ (p_1\ and\ p_2\ are\ 0) $$
$$A_m(\phi) = {\begin{bmatrix}
e^{-20h}& 0\\
\frac{1}{10}(1-e^{-20h})& 1
\end{bmatrix}}$$
$$W_c = [\Phi, \Gamma \Phi]$$
$$L = [0, 1]W_c^{-1}A_m(\Phi)$$
$By \ using \ Matlab, \ the \ result \ is$
$$L = {\begin{bmatrix}
-\frac{10h+e^{10h}-e^{20h}}{10h(e^{10h}-1)^2}\\
\frac{e^{10h}}{h(e^{10h}-1)}
\end{bmatrix}}$$

$$ u(k) = -Lx(k) =-{\begin{bmatrix}
-\frac{10h+e^{10h}-e^{20h}}{10h(e^{10h}-1)^2}\\
\frac{e^{10h}}{h(e^{10h}-1)}
\end{bmatrix}} x(k) $$





<font face="Time new roman" size = 6 align=center bold>c)</font> 

$$ u(0) = -Lx(0) = - \frac{5h+3e^{10h}-3e^{20h}}{5h(e^{10h}-1)^2}$$
$$ |u(0)|<1, h>0$$
$$ Can't\ be\ solve\ by\ hand\ and\ Matlab $$


<font face="Time new roman" size = 7 align=center bold>Q2</font> 

<font face="Time new roman" size = 6 align=center bold>a)</font> 

$$z(k) = \begin{bmatrix}
x(k)\\
v(k)
\end{bmatrix}$$

$$z(k+1) = \begin{bmatrix}
x(k+1)\\
v(k+1)
\end{bmatrix} = 
\begin{bmatrix}
0.5 & 1   & 1\\
0.2 & 0.8 & 0\\
0   & 0   & 1
\end{bmatrix}
\begin{bmatrix}
x(k)\\
v(k)
\end{bmatrix} +
\begin{bmatrix}
\Gamma  \\
0 
\end{bmatrix} u(k)
=
\begin{bmatrix}
\Phi   & \Phi_{xw}\\
0 & \Phi_w\\
\end{bmatrix} 
\begin{bmatrix}
x(k)\\
v(k)
\end{bmatrix} =
\begin{bmatrix}
1\\
1\\
0
\end{bmatrix}u(k)
$$
$$ y(k) = \begin{bmatrix}
1 & 0 & 0
\end{bmatrix} 
\begin{bmatrix}
x_1(k)\\
x_2(k)\\
v(k)
\end{bmatrix}
$$
$The \ state\ and\ v\ can\ be\ measured,\ assuming $
$$ u(k) = -L_c Z(k) = -Lx(k) - L_wv(k)$$
$$x(k+1) = \Phi x(k)- \Gamma(Lx(k)+L_wv(k))+ \Phi_v v(k) = (\Phi - \Gamma L)x(k) + (\Phi_w-\Gamma L_w)v(k)$$
$To\ cancel\ the\ disurbance\ we\ need\ (\Phi_w-\Gamma L_w)\ to\ be\ zero$
$$\Phi_w-\Gamma L_w= 
\begin{bmatrix}
1\\
0
\end{bmatrix} - 
\begin{bmatrix}
1\\
1
\end{bmatrix} L_w = 
\begin{bmatrix}
1-L_w\\
L_w
\end{bmatrix}\ Can\ not\ to\ be\ zero
$$

$So\ need\ to\ design\ L,\ let\ all\ the (\Phi - \Gamma L)\ poles\ to\ zero$

$$zI-(\Phi - \Gamma L) = 
\begin{bmatrix}
z-0.5+l_1 & l_1-1\\
l_2-0.2   & z-0.8+l_2
\end{bmatrix}
$$
$$ |zI-(\Phi - \Gamma L)| = z^2$$
$$l_1 = \frac{17}{22} = 0.77, l_2=\frac{29}{55} = 0.52$$
$Analyze the TF from v(k) to y(k):$
$$Y(z) = cX(z)\ ==>\ H_w(z) = c(zI-\Phi + \Gamma L)^{-1}(\Phi_{xw}-\Gamma L_w)v(z)$$
$Let\ H_w(1)\ =
\frac{5(5*l_w + 100z -100l_wz-28)}{500z^2-5z+1} = 0$
$$ l_w = \frac{72}{95} = 0.76$$
$$\therefore  u(k) = -
\begin{bmatrix}
0.77 & 0.52

\end{bmatrix}x(k)-0.76v(k)$$

<font face="Time new roman" size = 6 align=center bold>b)</font> 

$$ \because x(k) can\ be\ measured,\ when k\ =\ 0,\ we\ have:$$
$$x_1(1) = 0.5x_1(0) + x_2(0) + u(0) + v(0) $$
$$x_2(1) = 0.2x_1(0) + 0.8x_2(0) + 0.1u(0) $$
$$\because v(k)\ is\ constant$$
$$\therefore x(k)\  and\ v(k)\ can\ be\ measured$$
$$\therefore\ can\ design\ the\ same\ dead-beat\ controller\ as\ a)$$
$$  u(k) = -
\begin{bmatrix}
0.77 & 0.52
\end{bmatrix}x(k)-0.76v(k)$$

<font face="Time new roman" size = 6 align=center bold>c)</font> 

$The \ state\ and\ v\ can\ be\ measured,\ assuming $
$$ u(k) = -L_c Z(k) = -Lx(k) - L_wv(k)$$
$$\hat{z}(k+1) = 
\begin{bmatrix}
\Phi   & \Phi_{xw}\\
0 & \Phi_w\\
\end{bmatrix} \hat{z}(k)+
\begin{bmatrix}
\Gamma\\
0
\end{bmatrix}u(k)+K
\begin{bmatrix}
y(k)-\hat{y}(k)
\end{bmatrix} $$

$$ \hat{y}(k) = 
\begin{bmatrix}
c & 0
\end{bmatrix} \hat{z}(k)
\begin{bmatrix}
x_1(k)\\
x_2(k)\\
v(k)
\end{bmatrix} $$

$$ e(k) =z(k)- \hat{z}(k)$$
$$ e(k+1) =(\Phi - Kc)e(k)$$
$Place\ the\ poles\ to\ zero.\ The\ the\ K\ can\ be\ chosen.$
$$zI - (\Phi - Kc) =
\begin{bmatrix}
k1 + z - 1/2&      -1&    -1\\
    k2 - 1/5& z - 4/5&     0\\
          k3&       0& z - 1
\end{bmatrix}$$

$$det(zI - (\Phi - Kc)) = z^3$$
$$K = [2.3\ -2.36\ \ 5]^T$$
$The\ controller\ is:$
$$\therefore   u(k) = -
\begin{bmatrix}
0.77 & 0.52
\end{bmatrix}x(k)-0.76v(k)$$

<font face="Time new roman" size = 7 align=center bold>Q3</font> 

<font face="Time new roman" size = 6 align=center bold>a)</font> 

$First\ to\ build\ the\ reference\ model\ by\ using\ the\ overshoot\ and\ stabilization\ time:$
$$Magitude = e^{-\frac{\pi\zeta}{\sqrt{1-\zeta^2}}}<=5\%$$

In [1]:
import numpy as np
from sympy import *
zeta = symbols('zeta')
zeta_res = solve(exp(-pi*zeta/sqrt(1-zeta**2))-0.05, zeta)[0]
print('zeta:',zeta_res)
w0 =  4.6/6.9
print('w0:', w0)

zeta: 0.690106730559822
w0: 0.6666666666666666



$$\zeta = 0.69$$
$$t_s = \frac{4.6}{\zeta w_n} <= 10s $$
$$w_n = 0.67$$
$The\ reference\ model\ is:$
$$H_m(s) = \frac{w_n^2}{s^2 + 2\zeta w_n s+w_n^2} = \frac{w_n^2}{s^2 + 0.92s+0.43}$$


In [2]:
sampling = 0.1
w = w0*sqrt(1-zeta_res**2)
apha = exp(-zeta_res*w0 * sampling)
beta = cos(w*sampling)
gamma = sin(w*sampling)

b1 = 1-apha*(beta + zeta_res*w0*gamma/w)
b2 = apha**2 + apha*(zeta_res*w0*gamma/w - beta)
a1 = -2*apha*beta
a2 = apha**2
print(b1, b2, a1, a2)

0.00215480968809378 0.00208971642955746 -1.90784764358737 0.912092169705017


$The\ discrete-time\ trasfer\ function\ is:$
$$ H_m(s) = \frac{0.00215z + 0.0021}{z^2-1.9078z+0.9121}$$

<font face="Time new roman" size = 6 align=center bold>b)</font> 

$$ 1000 \ddot{y}(t) + 200\dot{y}(t) = u(t) $$
$$x_1(t) =y(t),\ x_2(t) = \dot{y}(t)$$
$$\left\{
    \begin{matrix} 
    \dot{x}_1 & = & x_2(t)  \\ 
    \dot{x}_2 & = & -0.2x_2 + 0.001u
    \end{matrix}
    \right. =
    \begin{bmatrix} 
    0     & 1 \\ 
    0     & -0.2
    \end{bmatrix}
    \begin{bmatrix} 
    x_1       \\ 
    x_2  
    \end{bmatrix} +
    \begin{bmatrix} 
    0       \\ 
    0.001 
    \end{bmatrix}u $$
$$ y(t) = \begin{bmatrix}  1&0 \end{bmatrix}x(t) $$
$$ A = \begin{bmatrix} 
    0     & 1 \\ 
    0     & -0.2
    \end{bmatrix}
$$
$$B = \begin{bmatrix} 
    0       \\ 
    0.001 
    \end{bmatrix}$$
$$C = \begin{bmatrix}  1&0 \end{bmatrix}$$
$The\ discrete-time\ mode\ is:$
$$ x(k+1) = \Phi x(k) + \Gamma u(k)  $$
$$ y(k) = Cx(k) $$


import scipy
A = [[0, 1], [0, -0.2]]
h = 0.1
Ah = np.dot(A,h)
dis_phi = scipy.linalg.expm(Ah)
print(dis_phi)

#t = sym('t');
#fun = expm(A*t);
#tao = double(int(fun,0,0.1))*B;
#tao


$$\Phi = e^{Ah} = 
\begin{bmatrix} 
    1  &  0.09900663    \\ 
    0  &  0.98019867
    \end{bmatrix}
    \ \ \ 
    \Gamma = \begin{bmatrix} 
    4.97  \times 10^{-6}    \\ 
    9.901 \times 10^{-5}  
    \end{bmatrix}
$$

<font face="Time new roman" size = 6 align=center bold>c)</font> 

$$W_c = 
\begin{bmatrix} 
\Gamma & \Phi \Gamma
\end{bmatrix} = 
    \begin{bmatrix} 
    0.0497  & 0.1477    \\ 
    0.9901  & 0.9705 
    \end{bmatrix} \ \ Rank(W_c) = 2
$$
$$u(k) = -Lx(k) = -l_1x_1(k)-l_2x_2(k)$$
$$x(k+1) = (\Phi - \Gamma L)x(k)$$
$Place\ the $
$ \Phi - \Gamma L \ poles\ to\ the\ target\ A_m(z) = z^2-1.9078z+0.9121 $
$$zI-(\Phi-\Gamma L)=
    \begin{bmatrix} 
    1-4.97 \times 10^{-6}l_1  & 0.099 - 4.97 \times 10^{-6}l_2    \\ 
    -9.901 \times 10^{-5}l_1  & 0.98 - 9.901 \times 10^{-5}l_2 
    \end{bmatrix}
$$
$$det(zI-(\Phi-\Gamma L))=z^2 + (-1.908+4.9668 \times 10^{-6} l_1 + 9.9007 \times 10^{-5} l_2)z + 0.9802 +4.9338 \times 10^{-6}l_1 - 9.9007 \times 10^{-5}l_2 $$
$$l_1 = 432.87, \ \ l_2 = 709.55$$
$$u_{fb}(k)=(-432.87, -709.55)x(k)$$

<font face="Time new roman" size = 6 align=center bold>d)</font> 

$$u_{ff}(k) = H_{ff}u_c(k)$$
$$x(k+1)=(\Phi-\Gamma L)x(k)+\Gamma H_{ff}u_c(k) $$
$$Z transfer: ZX(z) = (\Phi - \Gamma L)X(z) + \Gamma H_{ff}U_c(z)$$
$$X(z) = C(zI-\Phi + \Gamma L)^{-1}\Gamma H_ff U_C(z)$$
$$Y(z) = C(zI - \Phi +\Gamma L)^{-1}\Gamma H_{ff}$$
$$H_{ff} = \frac{B_m(z)}{B(z)} $$
$In\ question\ a)\ we\ can\ get\ \frac{B_m(z)}{A_m(z)}$
$$\frac{B_m(z)}{A_m(z)} = \frac{(0.00215z+0.002133)}{z^2-1.9078z+0.9121} $$
$$H(z) = \frac{B(z)}{A(z)} = C(zI-\Phi)^{-1}\Gamma = \frac{(4.9667z+4.9333)\times 10^{-6}}{z^2-1.9802z+0.9802} $$
$$H_{ff}(z) =  \frac{B_m(z)}{B(z)} = \frac{(0.00215z + 0.0021)\times 10^{6}}{4.9667z+4.9333}$$


<font face="Time new roman" size = 6 align=center bold>e)</font> 

$$W_o = 
\begin{bmatrix} 
\Gamma & \Phi \Gamma
\end{bmatrix} = 
    \begin{bmatrix} 
    1  & 0    \\ 
    1  & 0.099 
    \end{bmatrix} \ \ Rank(W_o) = 2
$$
$So\ x(k)\ can\ be\ estimated,\ design\ the\ feedback\ controller\ same\ as\ c)$
$$u_{fb}(k)=(-432.87, -709.55)x(k)$$
$$We have x(k+1) = \Phi x(k) +\Gamma u(k)$$
$$We have \hat{x}(k+1) = \Phi \hat{x}(k) +\Gamma u(k) + K(k(k) - \hat{y}(k))$$
$$let\ e(k) = x(k)- \hat{x}(t),\ e(k+1) = (\Phi - Kc)e(k)$$
$The\ dead-beat\ observer\ A_0(z) = z^2$
$$zI- (\Phi-Kc) = 
\begin{bmatrix} 
    z-1+k_1  &  -0.099  \\ 
    k_2      &  z-0.98
\end{bmatrix}
$$
$$det|zI - \Phi + Kc| = z^2 - 0.98z +(k_1-1)z - 0.98(k_1-1)+0.099k_2 = z^2$$
$$
\left\{
    \begin{matrix} 
    k_1 - 1-0.98   & = & 0  \\ 
    0.099k_2-0.98(k_1-1) & = & 0
    \end{matrix}
    \right.
$$
$$K = [1.98,\ k_2 = 9.7]^T$$
$$\therefore u_{fb}(k) = (-432.87, -709.55)\hat{x}(k)$$
$$\therefore u(k) = u_{fb} + u_{ff}$$
$$ It's\ possible\ to\ use\ two\ degree\ of\ freedom\ controller $$